code/game/machinery/vendors/vending.dm

```diff
@@ -0,0 +1,880 @@
+	. = ..()
+	var/build_inv = FALSE
+	if(!refill_canister)
+		build_inv = TRUE
+	else
+		component_parts = list()
+		var/obj/item/circuitboard/vendor/V = new
+		V.set_type(replacetext(initial(name), "\improper", ""))
+		component_parts += V
+		component_parts += new refill_canister
+		RefreshParts()
+
+	wires = new(src)
+	if(build_inv) //non-constructable vending machine
+		build_inventory(products, product_records)
+		build_inventory(contraband, hidden_records)
+		build_inventory(premium, coin_records)
+	for(var/datum/data/vending_product/R in (product_records + coin_records + hidden_records))
+		var/obj/item/I = R.product_path
+		var/pp = replacetext(replacetext("[R.product_path]", "/obj/item/", ""), "/", "-")
+		imagelist[pp] = "[icon2base64(icon(initial(I.icon), initial(I.icon_state), SOUTH, 1))]"
+
+	if(LAZYLEN(slogan_list))
+		// So not all machines speak at the exact same time.
+		// The first time this machine says something will be at slogantime + this random value,
+		// so if slogantime is 10 minutes, it will say it at somewhere between 10 and 20 minutes after the machine is created.
+		last_slogan = world.time + rand(0, slogan_delay)
+
+	if(account_database)
+		vendor_account = account_database.vendor_account
+
+	reconnect_database()
+	power_change()
+
+/obj/machinery/economy/vending/Destroy()
+	SStgui.close_uis(wires)
+	QDEL_NULL(wires)
+	QDEL_NULL(coin)
+	QDEL_NULL(inserted_item)
+	return ..()
+
+/obj/machinery/economy/vending/RefreshParts()         //Better would be to make constructable child
+	if(!component_parts)
+		return
+
+	product_records = list()
+	hidden_records = list()
+	coin_records = list()
+	if(refill_canister)
+		build_inventory(products, product_records, start_empty = TRUE)
+		build_inventory(contraband, hidden_records, start_empty = TRUE)
+		build_inventory(premium, coin_records, start_empty = TRUE)
+	for(var/obj/item/vending_refill/VR in component_parts)
+		restock(VR)
+
+/obj/machinery/economy/vending/update_overlays()
+	. = ..()
+	underlays.Cut()
+	if(panel_open)
+		. += "[icon_panel ? "[icon_panel]_panel" : "[icon_state]_panel"]"
+	if(icon_addon)
+		. += "[icon_addon]"
+	if((stat & (BROKEN|NOPOWER)) || force_no_power_icon_state)
+		. += "[icon_off ? "[icon_off]_off" : "[icon_state]_off"]"
+		if(stat & BROKEN)
+			. += "[icon_broken ? "[icon_broken]_broken" : "[icon_state]_broken"]"
+		return
+	if(light)
+		underlays += emissive_appearance(icon, "[icon_lightmask ? "[icon_lightmask]_lightmask" : "[icon_state]_off"]")
+
+/*
+ * Reimp, flash the screen on and off repeatedly.
+ */
+/obj/machinery/economy/vending/flicker()
+	if(flickering)
+		return FALSE
+
+	if(stat & (BROKEN|NOPOWER))
+		return FALSE
+
+	flickering = TRUE
+	INVOKE_ASYNC(src, TYPE_PROC_REF(/obj/machinery/economy/vending, flicker_event))
+
+	return TRUE
+
+/*
+ * Proc to be called by invoke_async in the above flicker() proc.
+ */
+/obj/machinery/economy/vending/proc/flicker_event()
+	var/amount = rand(5, 15)
```

Comment:
This is possibly a proc that is active over 16.5 seconds. Might be a bit long but that's a design thing for the most part

Reply:
Yeah, I think messing with this is a bit out of scope for the PR